### PR TITLE
overlay_1 example

### LINKED ISSUE #109: overlay_1
Description: Filled and unfilled contour with explicitly set contour levels

Script URL: https://www.ncl.ucar.edu/Applications/Scripts/overlay_1.ncl
Plot URL: https://www.ncl.ucar.edu/Applications/Images/overlay_1_lg.png

Category: Overlays

### PR DESCRIPTION
**Pull request by hCraker**
Closes #109 
image

### CODE REVIEW

**Comment by hCraker** [comment]
@clairefio @michaelavs I have called `gvutil.add_lat_lon_ticklabels(ax)` but the labels didn't show up. Have either of you had this issue before? I'm using the `ccrs.PlateCarree()` projection.

**Comment by michaelavs** [comment]
Looks awesome! Only minor change I would make is to increase the tick label sizes to better match the NCL plot, but thats all I could see so far!

**Comment by hCraker** [comment]
Here is the plot. Any suggestions?
image

File: Plots/Overlays/NCL_overlay_1.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+# Plot filled contour
+temp = t.plot.contourf(ax=ax, transform=ccrs.PlateCarree(), cmap=cmap, levels=t_lev, extend='neither', add_colorbar=False, add_labels=False)
+plt.colorbar(temp, ax=ax, ticks=np.arange(215, 270, 5), orientation='horizontal', pad=0.075)
+
+# Plot line contour
+wind = u.plot.contour(ax=ax, transform=ccrs.PlateCarree(), vmin=-5, vmax=35, levels=u_lev, colors='black', linewidths=0.5, add_labels=False)
+# Manually specify where contour labels will go using lat and lon coordiantes
+manual = [(-107, 52), (-79, 57), (-78, 47), (-103, 32), (-86, 23)]
+clabels = ax.clabel(wind, levels=u_lev, fmt='%d', inline=True, fontsize=10, manual=manual)
```
Comment (clairefio) [inline Plots/Overlays/NCL_overlay_1.py:66]: I noticed that in the original NCL plot, the contour labels have white backgrounds. I tried to implement the method found here:
https://stackoverflow.com/questions/15957774/how-to-set-the-background-color-of-contour-labels-in-matplotlib
but I couldn't seem to get it to work, so maybe this is not a possibility.

File: Plots/Overlays/NCL_overlay_1.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+print(clabels)
```
Comment (clairefio) [inline Plots/Overlays/NCL_overlay_1.py:67]: unnecessary print statement

File: Plots/Overlays/NCL_overlay_1.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        horizontalalignment='right',
+        transform=ax.transAxes,
+        bbox=dict(boxstyle='square, pad=0.25', facecolor='white', edgecolor='black'))
+
+gvutil.set_titles_and_labels(ax, maintitle=r"$\bf{T/U @500hPa}$", lefttitle=t.long_name, righttitle=t.units)
+ax.text(0, 1.01, u.long_name, transform=ax.transAxes)
+ax.text(0.97, 1.01, u.units, transform=ax.transAxes)
+
+# Use geocat.viz.util convenience function to make plots look like NCL plots by using latitude, longitude tick labels
+gvutil.add_lat_lon_ticklabels(ax)
```
Comment (clairefio) [inline Plots/Overlays/NCL_overlay_1.py:81]: I would take the degree symbols out to match the NCL graph.
```suggestion
gvutil.add_lat_lon_ticklabels(ax)
ax.yaxis.set_major_formatter(LatitudeFormatter(degree_symbol= ''))
ax.xaxis.set_major_formatter(LongitudeFormatter(degree_symbol=''))
```

File: Plots/Overlays/NCL_overlay_1.py (hunk 1 above)
Comment (hCraker) [reply Plots/Overlays/NCL_overlay_1.py:67]: I looked at that same post before opening this PR and couldn't get the label backgrounds to be white. I'm going to look for another solution, but it very well may not be easy to do with matplotlib.

**Comment by hCraker** [comment]
This is what I have so far. I am still unable to get the contour labels to have white backgrounds. Any suggestions are appreciated.
image

File: Plots/Overlays/NCL_overlay_1.py [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+plt.colorbar(temp, ax=ax, ticks=np.arange(215, 270, 5),
+             orientation='horizontal', pad=0.075)
+
+# Plot line contour
+wind = u.plot.contour(ax=ax, transform=ccrs.PlateCarree(), vmin=-5, vmax=35,
+                      levels=u_lev, colors='black', linewidths=0.5,
+                      add_labels=False)
+# Manually specify where contour labels will go using lat and lon coordiantes
+manual = [(-107, 52), (-79, 57), (-78, 47), (-103, 32), (-86, 23)]
+clabels = ax.clabel(wind, levels=u_lev, fmt='%d', inline=True, fontsize=10,
```
Comment (michaelavs) [inline Plots/Overlays/NCL_overlay_1.py:70]: When I ran the code on my side, it didn't like that you defined `levels=u_lev` but was fine once I removed the "levels=" part. I'm also looking over your question about creating a white background for the contour line labels and will let you know if I find anything useful!

Comment (hCraker) [reply Plots/Overlays/NCL_overlay_1.py:70]: Noted. I'll fix it, thanks!

**Comment by hCraker** [comment]
Update on label backgrounds:
`matplotlib.axes.Axes.clabel` is what I'm using to format the contour labels. It it supposed to return a list of `Text` instances for each label, but it is returning `NoneType`. I'm not sure why this is. I searched through the matplotlib documentation and GitHub to see if this had been reported as a bug, but I didn't see an issue for it. I found code where the return value of `clabel()` is actually used and can't possibly be `NoneType` so apparently something is being returned for those contributors. I have tried seeing if `xarray.plot.contour` and `plt.contour` interacted dfferently with `clabel()`, but the function returns `NoneType` with both.

All of the suggested methods to color the background of the contour labels requires the return value of `clabel()`. Can anyone else print out the return value of `clabel()` and tell me what it is? I'm wondering if something is wrong with my environment. @clairefio @michaelavs

**Comment by michaelavs** [comment]
> Update on label backgrounds:
> `matplotlib.axes.Axes.clabel` is what I'm using to format the contour labels. It it supposed to return a list of `Text` instances for each label, but it is returning `NoneType`. I'm not sure why this is. I searched through the matplotlib documentation and GitHub to see if this had been reported as a bug, but I didn't see an issue for it. I found code where the return value of `clabel()` is actually used and can't possibly be `NoneType` so apparently something is being returned for those contributors. I have tried seeing if `xarray.plot.contour` and `plt.contour` interacted dfferently with `clabel()`, but the function returns `NoneType` with both.
> 
> All of the suggested methods to color the background of the contour labels requires the return value of `clabel()`. Can anyone else print out the return value of `clabel()` and tell me what it is? I'm wondering if something is wrong with my environment. @clairefio @michaelavs

I just did a print out of clabels and it is returning NoneType on my side as well. I think when I was looking into this the other day, I was seeing the same thing, if not that there were other functions that could be called after clabels and they would allow for the white box we are aiming for. However, when I would directly copy those codes in to experiment, it would give an error but (in theory) should not have. I wonder if there has been an update to this function at all and that may be why it isn't working the way it is expected to.

**Comment by hCraker** [comment]
Note for later: When this [issue](https://github.com/NCAR/geocat-viz/issues/7) is resolved, the code for this example should be updated to use the modified utility function.

**Comment by hCraker** [comment]
> I just did a print out of clabels and it is returning NoneType on my side as well. I think when I was looking into this the other day, I was seeing the same thing, if not that there were other functions that could be called after clabels and they would allow for the white box we are aiming for. However, when I would directly copy those codes in to experiment, it would give an error but (in theory) should not have. I wonder if there has been an update to this function at all and that may be why it isn't working the way it is expected to.

I checked the version of matplotlib in my environment and it is 3.2.1 which is the most recent stable release. The documentation for 3.2.1 says that `clabel` should return a list of `Text` instances, so I don't think there has been an update of sorts. I wonder if it's a bug in matplotlib.

File: Plots/Overlays/NCL_overlay_1.py (hunk 4 above)
Comment (kmpaul) [reply Plots/Overlays/NCL_overlay_1.py:70]: It works for me.

**Comment by kmpaul** [comment]
@michaelavs @hCraker I can confirm (with a freshly created conda env) that the `clabels` call returns `None`.

**Comment by hCraker** [comment]
> Why was `NCL_overlay_11b.py` deleted?

Whoops that was a mistake. I accidentally added `NCL_overlay_11b.py` to be a tracked file in my branch and I must have deleted the file instead of removing it from git tracking. I can fix that really quick.

File: Plots/Overlays/NCL_overlay_1.py (hunk 4 above)
Comment (hCraker) [reply Plots/Overlays/NCL_overlay_1.py:70]: Hmm interesting. I removed the keyword already and it still works for me. I have no idea why some of us would be getting warnings an others aren't given that we are all using the environments made by the `condaa_environment.yml` file

**Comment by kmpaul** [comment]
I've looked at the `matplotlib` code and it definitely shows the return value of the `clabel` function should be a `matplotlib.cbook.silent_list`.  I'm not sure why it returns `None`, but the fact that it is not a *true* `list` type is suspicious.

If you need the *untainted* list of labels, though, I verified that you can get it from the `labelTexts` attribute of the `ContourSet` object.  In this case, you can change the code from:

```python
clabels = ax.clabel(wind, ...)
```

to

```python
ax.clabel(wind, ...)
clabels = wind.labelTexts
```

File: Plots/Overlays/NCL_overlay_1.py (hunk 4 above)
Comment (hCraker) [reply Plots/Overlays/NCL_overlay_1.py]: @michaelavs Now that I have changed it, are you still getting a warning?

**Comment by hCraker** [comment]
I believe I'm nearly finished. Here's what it looks like.
image

File: Plots/Overlays/NCL_overlay_1.py [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+wind = u.plot.contour(ax=ax, transform=ccrs.PlateCarree(), vmin=-5, vmax=35,
+                      levels=u_lev, colors='black', linewidths=0.5,
+                      add_labels=False)
+# Manually specify where contour labels will go using lat and lon coordiantes
+manual = [(-107, 52), (-79, 57), (-78, 47), (-103, 32), (-86, 23)]
+ax.clabel(wind, u_lev, fmt='%d', inline=True, fontsize=10,
+          manual=manual)
+# Set label backgrounds white
+clabels = wind.labelTexts
+[txt.set_bbox(dict(facecolor='white', edgecolor='none', pad=2)) for txt in clabels]
```
Comment (kmpaul) [inline Plots/Overlays/NCL_overlay_1.py:75]: This is nice!  Good solution.

**Comment by hCraker** [comment]
> Great job Rose! I was wondering if there was a way to make adding white text boxes more intuitive to future users by writing a helper function for geocat-viz that does this, but hopefully the solution is simple enough that people looking at the gallery can find it.
> 
> That makes me think that maybe we want to add a tagging system to the gallery so that if someone searched, "How to have a white background box on contour labels," this example would show up. Perhaps the simplest way of implementing this is by having text captions or figure labels on all of our plots that point out some key features. Overall I think it would increase the use and effectiveness of the gallery. Thoughts? @clyne @kmpaul

I 100% agree that we should have a tagging system or something similar. A few solutions come to mind. We can make the comments describing what each example illustrates more complete to make searching for the right example simpler. 

Alternatively, we can model what the matplotlib documentation does. Those pages link examples that use the function described on that page. We could do something similar and add to the gallery pages that discuss how to format titles and labels, add lat/lon ticks to a projection, etc and then link relevant examples.

**Comment by clyne** [comment]
Great ideas, both!

**Comment by hCraker** [comment]
> I 100% agree that we should have a tagging system or something similar. A few solutions come to mind. We can make the comments describing what each example illustrates more complete to make searching for the right example simpler.
> 
> Alternatively, we can model what the matplotlib documentation does. Those pages link examples that use the function described on that page. We could do something similar and add to the gallery pages that discuss how to format titles and labels, add lat/lon ticks to a projection, etc and then link relevant examples.

@clyne Should I make adding pages for specific plot elements an issue in this repo?

File: Plots/Overlays/NCL_overlay_1.py (hunk 5 above)
Comment (kmpaul) [inline Plots/Overlays/NCL_overlay_1.py:75]: It doesn't look like `clabels` is used anywhere else, so you could (at the risk of the line getting too long) make this a one-liner:

```suggestion
[txt.set_bbox(dict(facecolor='white', edgecolor='none', pad=2)) for txt in wind.labelTexts]
```

Arguably, that is more "pythonic."

**Comment by clyne** [comment]
> > I 100% agree that we should have a tagging system or something similar. A few solutions come to mind. We can make the comments describing what each example illustrates more complete to make searching for the right example simpler.
> > Alternatively, we can model what the matplotlib documentation does. Those pages link examples that use the function described on that page. We could do something similar and add to the gallery pages that discuss how to format titles and labels, add lat/lon ticks to a projection, etc and then link relevant examples.
> 
> @clyne Should I make adding pages for specific plot elements an issue in this repo?

Sounds good to me.

File: Plots/Overlays/NCL_overlay_1.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
+"""
+NCL_overlay_1.py
+===============
+This script illustrates the following concepts:
+   - Overlaying line contours on filled contours
+   - Explicitly setting contour levels
+   - Adding custom formatted contour labels
+   - Mannually selecting where contour labels will be drawn
```
Comment (erogluorhan) [inline Plots/Overlays/NCL_overlay_1.py:8]: Typo "Manually"

File: Plots/Overlays/NCL_overlay_1.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+import cartopy.crs as ccrs
+import cartopy.feature as cfeature
+from cartopy.mpl.gridliner import LongitudeFormatter, LatitudeFormatter
+
+import geocat.datafiles as gdf
+from geocat.viz import util as gvutil
+from geocat.viz import cmaps as gvcmaps
+
+###############################################################################
+# Read in data:
```
Comment (erogluorhan) [inline Plots/Overlays/NCL_overlay_1.py:31]: Leave a line break between two comments

File: Plots/Overlays/NCL_overlay_1.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+[txt.set_bbox(dict(facecolor='white', edgecolor='none', pad=2)) for txt in wind.labelTexts]
+
+# Add lower text box
+ax.text(1, -0.3, "CONTOUR FROM -5 TO 35 BY 5",
+        horizontalalignment='right',
+        transform=ax.transAxes,
+        bbox=dict(boxstyle='square, pad=0.25', facecolor='white',
+                  edgecolor='black'))
+
+gvutil.set_titles_and_labels(ax, maintitle=r"$\bf{T/U @500hPa}$",
```
Comment (erogluorhan) [inline Plots/Overlays/NCL_overlay_1.py:83]: Put a descriptive comment

File: Plots/Overlays/NCL_overlay_1.py [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+                             lefttitle=t.long_name, righttitle=t.units)
+ax.text(0, 1.01, u.long_name, transform=ax.transAxes)
+ax.text(0.97, 1.01, u.units, transform=ax.transAxes)
+
+# Use geocat.viz.util convenience function to make plots look like NCL plots by
+# using latitude, longitude tick labels
+gvutil.add_lat_lon_ticklabels(ax)
+ax.yaxis.set_major_formatter(LatitudeFormatter(degree_symbol=''))
+ax.xaxis.set_major_formatter(LongitudeFormatter(degree_symbol=''))
+gvutil.add_major_minor_ticks(ax, x_minor_per_major=3, y_minor_per_major=5,
```
Comment (erogluorhan) [inline Plots/Overlays/NCL_overlay_1.py:93]: Separate this utility function from above lines and out its own comment.

**Pull request merged**